content/blog/2020-12-20-bril-algebraic-data-types/index.md

```diff
@@ -0,0 +1,538 @@
+
+Here is a sample Bril program that uses tuples:
+
+```
+@main() {
+  one: int = const 1;
+  two: int = const 2;
+  tuple: product<int, int> = pack one two;
+  one_again = unpack tuple 0;
+}
+```
+
+### Sum types
+
+#### `construct`
+
+The `construct` instruction is used to construct sum types. It takes
+as its sole argument the value to use. The semantics of the
+instruction dictate that constructor to use is determined
+automatically by examining the runtime type of the argument.
+
+The destination must have a sum type and the type of the argument must
+correspond to one of the constructors of that sum type.
+
+The JSON format of the `construct` instruction is as follows:
+
+```json
+{
+    "op": "construct",
+    "args": ["a"],
+    "dest": "d",
+    "type": {"sum": [t1, ..., tn]}
+}
+```
+
+The text format of the `construct` instruction is as follows:
+
+```
+d: sum<t1, ..., tn> = construct a
+```
+
+In theory, it could have been possible to omit `construct` entirely if
+there were a sum type subtyping rule in place, but such a rule would
+probably have made things more, not less, confusing.
+
+#### `destruct`
+
+The `destruct` instruction is used to execute different code based on
+the constructor used to create a sum-typed value. It takes a sum-typed
+value as an argument and a number of labels, one for each constructor
+in the sum type.
+
+The destination of the instruction is a variable name with a sum
+type. The semantics of the instruction dictate that the control flow
+jumps to the correct branch based on the argument and that the
+destination value gets assigned the value within the constructor.
+
+The destination will in fact not have the same sum type as the
+argument; it will have the type carried by one of the constructors,
+which is different for each label. This is confusing because the
+destination is annotated with a type that it cannot have, but I chose
+this approach because the Bril interpreter needs to know the available
+constructors in order to select the correct label to jump to.
+
+The JSON format of the `destruct` instruction is as follows:
+
+```json
+{
+    "op": "destruct",
+    "args": ["a"],
+    "labels": ["l1", ..., "ln"],
+    "dest": "d",
+    "type": {"product": [t1, ..., tn]}
+}
+```
+
+The text format of the `destruct` instruction is as follows:
+
+```
+d: sum<t1, ..., tn> = destruct v .l1 ... .ln
+```
+
+A common use case is some way of representing an empty, or unit,
+value. For this it is recommended to use the empty product type,
+i.e. `product<>`. For example, to represent an optional `int`, one
+could use the following type: `sum<int, product<>>`.
+
+I debated whether to even include a powerful instruction like
+`destruct` at all. I could have instead offered an instruction to get
+the index of the constructor used for a value and another instruction
+for casting a sum type down to one of the constructor-carried
+types, using a series of branch instructions to replicate the behavior
+of `destruct`.
```

Comment:
Aha, you had the same idea as me. :smiley: Leaving my comment in there above for posterity.